retrieval/scrape.go

```diff
@@ -0,0 +1,294 @@
+	reportAppender storage.SampleAppender
+
+	// A function to calculate an initial offset for a given
+	// scrape interval.
+	offset func(time.Duration) time.Duration
+
+	stopc, done chan struct{}
+	cancel      func()
+	mtx         sync.RWMutex
+}
+
+func newScrapeLoop(app storage.SampleAppender, target *Target, state *TargetStatus) loop {
+	return &scrapeLoop{
+		scraper:        target,
+		appender:       target.wrapAppender(app, true),
+		reportAppender: target.wrapAppender(app, false),
+		state:          state,
+		offset:         target.offset,
+	}
+}
+
+func (sl *scrapeLoop) active() bool {
+	return sl.cancel != nil
+}
+
+func (sl *scrapeLoop) run(ctx context.Context, interval, timeout time.Duration, errc chan<- error) {
+	if sl.active() {
+		return
+	}
+
+	// If no offset function is provided, default to starting immediately.
+	if sl.offset == nil {
+		sl.offset = func(time.Duration) time.Duration { return 0 }
+	}
+
+	ctx, sl.cancel = context.WithCancel(ctx)
+
+	sl.stopc = make(chan struct{})
+	sl.done = make(chan struct{})
+
+	defer func() {
+		close(sl.done)
+		sl.cancel = nil
+	}()
+
+	select {
+	case <-time.After(sl.offset(interval)):
+		// Continue after a scraping offset.
+	case <-ctx.Done():
+		return
+	case <-sl.stopc:
+		return
+	}
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		start := time.Now()
+
+		err := sl.scrape(ctx, timeout)
+		if err != nil {
+			log.Debugf("Scraping error: %s", err)
+		}
+
+		sl.report(start, time.Since(start), err)
+		if errc != nil {
+			errc <- err
+		}
+
+		// First check without the ticker to not initiate
+		// another scrape if termination was requested.
+		select {
+		case <-sl.stopc:
+			return
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-sl.stopc:
+			return
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// kill terminates the scraping loop immediately.
+func (sl *scrapeLoop) kill() {
+	if !sl.active() {
+		return
+	}
+	sl.cancel()
+}
+
+// stop finishes any pending scrapes and then kills the loop.
+func (sl *scrapeLoop) stop() {
+	if !sl.active() {
+		return
+	}
+	close(sl.stopc)
+	<-sl.done
+
+	sl.kill()
+}
+
+// scrape executes a single Scrape and appends the retrieved samples.
+func (sl *scrapeLoop) scrape(ctx context.Context, timeout time.Duration) error {
+	ch := make(chan *model.Sample, ingestedSamplesCap)
+
+	// Receive and append all the samples retrieved by the scraper.
+	go func() {
+		var (
+			smpl *model.Sample
+			ok   bool
+		)
+		for {
+			select {
+			case smpl, ok = <-ch:
+				if !ok {
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+			// TODO(fabxc): Change the SampleAppender interface to return an error
+			// so we can proceed based on the status and don't leak goroutines trying
+			// to append a single sample after dropping all the other ones.
+			sl.appender.Append(smpl)
+		}
+	}()
+
+	ctx, _ = context.WithTimeout(ctx, timeout)
+
+	return sl.scraper.Scrape(ctx, ch)
```

Comment:
Valid point regarding performance, don't think this is a good solution though.
The point of this PR is fixing all the logic leaks, i.e. scrape loops worrying about wrapping samples appenders with relabeling, target managers worrying about scraping, etc.

Similarly a `Scrape()` function shouldn't know such a thing as a `SampleAppender`.

Making the channel `chan<- []*model.Sample` should allow us to maintain separation here without significant performance impact.